RFC-0001-error_messages_from_resource_bundle.md

```diff
@@ -0,0 +1,81 @@
+# **RFC1 for Presto**
+
+See [CONTRIBUTING.md](CONTRIBUTING.md) for instructions on creating your RFC and the process surrounding it.
+
+## Loading error messages from resource bundles / localization of error messages
+
+Proposers
+
+* Elbin Pallimalil
+
+## [Related Issues]
+
+https://github.com/prestodb/presto/issues/21199
+
+## Summary
+
+Move all the inline error messages to resource bundles. Take advantage of this change to serve client locale specific error messages from Presto.
+
+## Background
+
+Presto now has inline error messages written in multiple locations. If we need to review the error messages we don't have a central location to review all the messages. We can move the inline error messages to resource files to keep them in a central location.
+
+When we are moving the error messages to resource files, we can also load locale specific error messages during server startup so that we can serve client locale specific error messages from Presto. 
+
+## Proposed Implementation
+
+### Presto implementation
+
+ 1. Create a default Messages.properties resource under `src/main/resources/error` folder in presto-main module 
+ 2. Add all error messages used in Presto code in Messages.properties 
+ 3. Deprecate inline error messages in Presto code and load message from Messages.properties 
+ 4. Only load locale specific error messages if server property `error.i18n.enabled` is true. By default it's value will be false.
+ 4. Check if there are Messages files for different locales in `etc/resources/error` folder during Presto startup. If yes load those locale specific resources. 
+ 5. When loading messages from resources, load from the locale specific resource. The locale to be used should be determined from the client session. 
+ 6. Each plugin can define additional error messages under `src/main/resources/error` folder in the corresponding plugin module. 
+ 7. These resource files from plugins will be read on server startup and a combined resource file for each locale will be created. 
+ 8. On server startup, read locale specific Messages.properties files for connector from a configurable path.
+ 9. Read the configurable path from `error.resources.location` connector property. The root will be `plugin/<connector-name>` and the default value will be `resources/error` 
+ 10. Load localized error messages in `QueuedStatementResource.toQueryError` method by getting the client locale from `sessionContext.language`
+ 11. Warn when the default resource bundle entries are not the same as the translated entries 
+ 12. When a translation is not available, a locale message is transcribed that explains an error message is not available, and then prints the message in the default locale (English).  That makes it embarrassing to the deployer of Presto, but less bad than preventing debugging. 
+ 13. Presto exceptions in query logs contain both the default and the locale-specific exception messages 
+ 14. Compile time validation of error message keys. 
```

Comment:
Will this work for connectors?

Reply:
The connectors will need to implement unit test cases that validate error keys. A sample unit test might look like below.

<img width="870" alt="image" src="https://github.com/prestodb/rfcs/assets/148545490/4a46ede5-d599-40f1-b39c-082fb7fbdd65">

Reply:
<img width="935" alt="Screenshot 2024-02-21 at 12 04 02 PM" src="https://github.com/prestodb/rfcs/assets/148545490/4756d4e7-d995-4e6d-88dc-9fa717abc27a">
Added the validation of connector keys as point 16.

Reply:
<img width="1051" alt="image" src="https://github.com/prestodb/rfcs/assets/148545490/fbc3d1bc-2d49-44ad-9ce5-3026fd83e354">
These points describe how connectors will load the error message .properties file.

Reply:
Is the unit test something that must be written manually for each connector?

Reply:
As of now, the design is such that each connector needs to define this unit test.